### ISSUE TITLE
Extend the language to be able to convert an action to a temporal formula

### ISSUE THREAD

**Issue by konnov** [informalsystems/quint#1236]
This has been raised in the [discussion](https://github.com/informalsystems/quint/pull/1178#discussion_r1384749374).

Like in the discussion, we would like to produce a temporal formula out of `init`, which would be evaluated against the current state. For example:

```bluespec
action init = all { x = 1, y = 2 }

// the two following formulas should be equivalent
temporal myFormula1 = temporal(init)
temporal myFormula2 = x == 1 and y == 2
```

**Comment by bugarela** [comment]
Should this result in an error?
```bluespec
action A = all { x == 0, x' = 1 }

temporal F = temporal(A)
```

**Comment by konnov** [comment]
I guess so. But we should think about this operator. Perhaps, we could introduce `pre(A)` and `post(A)` for the pre- and postcondition. I actually like that. In this case, `post(init)` would be the post-condition, whereas `pre(A)` would be just `true`.

**Comment by bugarela** [comment]
I like that! `pre()` seems like what I wanted `enabled` for. Would be great to have that even outside the temporal mode, if possible.

**Comment by thpani** [comment]
> In this case, `post(init)` would be the post-condition, whereas `pre(A)` would be just `true`.
> 
> @thpani, @Kukovec, do you think this would be the right treatment of pre/post? Or am I missing anything?

Can you explain on the example what `pre(A)` and `post(A)` would evaluate to?
For me, `pre` and `post` invoke Dijkstra's predicate transformers, but I think you're trying to do something different here.

**Comment by bugarela** [comment]
In my understanding, `pre(A)` would result in `x == 0` and `post(A)` in `x == 1`.

And if we have something like
```
action B = x' = x + 1
```

`pre(B)` is true, but `post(B)` should give a failure, I think.

**Comment by Kukovec** [comment]
> pre(B) is true, but post(B) should give a failure, I think.

That is very unintuitive. If anything, I'd expect `post(B) = x == N` for whatever value `N = x + 1`  fits.

**Comment by Kukovec** [comment]
But if you want a decomposition invariant, the only way I see it is if it is always the case that 
```
A == pre(A) /\ post(A)
```
in which case `post(A)` is a conjunction of only primed equalities, e.g.
```
action A = and {
  x > N,
  x' = x,
  y' = x ^ 2
}
```
would give
```
pre(A) = x > N
post(A) = x' = x and y' = x^2
```

**Comment by bugarela** [comment]
Yeah, that makes more sense. I was thinking about `post` as something equivalent to what we were talking about for a `temporal` operator (this issue). But it might be better to have the three of them: `pre`, `post` and `temporal`.

**Comment by konnov** [comment]
> I like that! `pre()` seems like what I wanted `enabled` for. Would be great to have that even outside the temporal mode, if possible.

Oh, that's true. Actually, `pre` seems to be exactly like `ENABLED` under the restrictions on the actions that we have.

**Comment by konnov** [comment]
> > In this case, `post(init)` would be the post-condition, whereas `pre(A)` would be just `true`.
> > @thpani, @Kukovec, do you think this would be the right treatment of pre/post? Or am I missing anything?
> 
> Can you explain on the example what `pre(A)` and `post(A)` would evaluate to? For me, `pre` and `post` invoke Dijkstra's predicate transformers, but I think you're trying to do something different here.

Yeah, I am not sure about it myself. Let's consider the following example (made a bit more complicated on purpose):

```bluespec
action collatz = any {
  all {
    x % 2 == 0,
    x' = x / 2
  },
  all {
    x % 2 != 0,
    x' = 3 * x + 1
  }
}
```

So in my imagination it would be like that:

```bluespec
// this is obviously equivalent to `true`,
// but this is how we could compute it without thinking, in a pure mechanic way:
pre(collatz) == or {
  x % 2 == 0,
  x % 2 != 0
}

post(collatz) == or {
  and { x % 2 == 0, next(x) = x' = x / 2 },
  and { x % 2 != 0, next(x) = x' = 3 * x + 1 },
}
```

Now that I have written down this example, `pre` seems to match my intuition of the weakest precondition for `wp(collatz, true)`. This is probably exactly `ENABLED` in Quint, as @bugarela noted. However, `post` is definitely not any kind of a postcondition. It is just the Quint action `collatz` written as a relation, which would be just an action in TLA+ :)

Going back to the very first example, would not we be able to capture the result of `init` with `sp(init, true)`, which would gives us `x == 1 and y == 2`?

**Comment by bugarela** [comment]
That example makes sense to me!

I don't know what you mean with `wp` and `sp` tho

**Comment by Kukovec** [comment]
"weakest precondition" and "strongest postcondition"

**Comment by konnov** [comment]
> > Going back to the very first example, would not we be able to capture the result of `init` with `sp(init, true)`, which would give us `x == 1 and y == 2`?
> 
> Exactly.
> 
> The longer example makes sense, though I don't understand where I'd use `post`. It's just a normal form then, right?

Yeah, I am not sure about it either. Normally, one would need an action in TLA+ to express weak fairness and strong fairness. So maybe there is a need. We should not call it `post` though. It is simply an action written as a temporal formula.

**Comment by konnov** [comment]
Also, we have established that we need `sp(init, true)` to transform `init` into a temporal formula. Do we want to introduce an operator for exactly `sp(A, true)`, or would it make sense to have `sp(A, F)` for an arbitrary formula `F`?
